### PR TITLE
Replicate stack gan

### PR DESCRIPTION
**Pull request by Devin-Taylor**
- disabled biases in Conv layers
- added wrong images to the dataset and to loss function of the discriminator
- fix KL loss function
- crop images to 75% object to image ratio

### CODE REVIEW

File: models/generators.py [hunk 1]
```diff
@@ -101,6 +97,8 @@ def __init__(self, img_size, kernel_size, num_filters,
         assert num_output_channels == 3 or num_output_channels == 1, \
             f'The number of output channels must be 2 or 1. Found {num_output_channels}'
 
+        self.optimiser = tf.keras.optimizers.Adam(lr, beta_1=0.5)
+
```
Comment (alecokas) [inline models/generators.py:101]: Why can't this stay in the parent class?

File: models/generators.py [hunk 2]
```diff
@@ -150,24 +150,21 @@ def call(self, embedding, noise):
...
-        )
-        loss = tf.reduce_mean(loss) + kl_coeff * self.kl_loss(mean, log_sigma)
+        kl_coeff = 1
+        loss = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(labels=tf.ones_like(predictions_on_fake), logits=predictions_on_fake))
+        loss = loss + kl_coeff * self.kl_loss(mean, log_sigma)
         return loss
 
     def kl_loss(self, mean, log_sigma):
-        loss = -log_sigma + .5 * (-1 + tf.exp(2. * log_sigma) + tf.square(mean))
+        loss = -.5 * log_sigma + .5 * (-1 + tf.exp(2. * log_sigma) + tf.math.square(mean))
```
Comment (alecokas) [inline models/generators.py:167]: You could have just moved this into the bracket but it doesn't matter :stuck_out_tongue_closed_eyes:

File: trainers/text_to_image_trainer.py [hunk 3]
```diff
@@ -36,26 +42,36 @@ def train_epoch(self, train_loader, epoch_num):
...
                         sample=sample,
                         index=i,
                         embedding_size=1024,
-                        num_embeddings_to_sample=self.num_embeddings
+                        num_embeddings_to_sample=self.num_samples,
+                        include_wrong=True
                     )
-                    img = np.asarray(img)
+                    img = np.asarray(img, dtype='float32')
+                    wrong_img = np.asarray(wrong_img, dtype='float32')
```
Comment (alecokas) [inline trainers/text_to_image_trainer.py:60]: Yup ok it is here

File: models/generators.py [hunk 4]
```diff
@@ -150,24 +150,21 @@ def call(self, embedding, noise):
...
         """
-        kl_coeff = 2
-        loss = tf.nn.sigmoid_cross_entropy_with_logits(
-            labels=tf.ones_like(predictions_on_fake),
-            logits=predictions_on_fake
-        )
-        loss = tf.reduce_mean(loss) + kl_coeff * self.kl_loss(mean, log_sigma)
+        kl_coeff = 1
+        loss = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(labels=tf.ones_like(predictions_on_fake), logits=predictions_on_fake))
+        loss = loss + kl_coeff * self.kl_loss(mean, log_sigma)
```
Comment (alecokas) [inline models/generators.py:163]: Really not a fan of short lines I see haha

File: models/discriminators.py [hunk 5]
```diff
@@ -45,10 +37,13 @@ def __init__(self, img_size, kernel_size, num_filters, lr, w_init, bn_init):
...
-        self.conv_1 = Conv2D(
-            filters=64, kernel_size=(4, 4), strides=(2, 2), padding='same',
-            kernel_initializer=self.w_init
-        )
+        self.w_init = tf.random_normal_initializer(stddev=0.02)
+        self.bn_init = tf.random_normal_initializer(1., 0.02)
+
+        self.img_size = img_size
+        num_channels = self.img_size[0]
+        self.optimiser = tf.keras.optimizers.Adam(lr, beta_1=0.5)
```
Comment (alecokas) [inline models/discriminators.py:45]: This can stay in the parent class surely?
